lib/rpmsg/rpmsg_virtio.c

```diff
@@ -396,6 +468,8 @@ static void rpmsg_virtio_rx_callback(struct virtqueue *vq)
 	metal_mutex_release(&rdev->lock);
 
 	while (rp_hdr) {
+		rp_hdr->reserved = 0;
```

Comment:
There is a bug here. Consider the following scenario:-
a. Master sends a message to the remote.
b. The remote callback function is called.
c. Remote user holds the buffer but also decides to release the buffer before callback returns.

The bug is that in the above scenario, the rpmsg_virtio_release_rx_buffer function always free buffer with index zero because of this line.

This is a valid scenario, and I think the above line should be changed to:-
rp_hdr->reserved = idx;

And the function rpmsg_virtio_hold_rx_buffer should not overwrite this information by taking a bitwise OR as shown below:-
-	rp_hdr->reserved = RPMSG_BUF_HELD;
+	rp_hdr->reserved |= RPMSG_BUF_HELD;

Reply:
@kr-satish good finding, I will fix it in the update.